saleor/graphql/discount/types/sales.py

```diff
@@ -127,81 +125,112 @@ class Meta:
             + " Use `Promotion` type instead."
         )
         interfaces = [relay.Node, ObjectWithMetadata]
-        model = models.Sale
+        model = models.Promotion
+        doc_category = DOC_CATEGORY_DISCOUNTS
 
     @staticmethod
-    def resolve_created(root: models.Sale, _info: ResolveInfo):
-        return root.created_at
+    def resolve_created(root: ChannelContext[models.Promotion], _info: ResolveInfo):
+        return root.node.created_at
+
+    @staticmethod
+    def resolve_type(root: ChannelContext[models.Promotion], info: ResolveInfo):
+        if (
+            rules := PromotionRulesByPromotionIdLoader(info.context)
+            .load(root.node.id)
+            .get()
+        ):
+            return rules[0].reward_value_type
 
     @staticmethod
     def resolve_categories(
-        root: ChannelContext[models.Sale], info: ResolveInfo, **kwargs
+        root: ChannelContext[models.Promotion], info: ResolveInfo, **kwargs
     ):
-        qs = root.node.categories.all()
-        return create_connection_slice(qs, info, kwargs, CategoryCountableConnection)
+        if (
+            predicates := PredicateByPromotionIdLoader(info.context)
+            .load(root.node.id)
+            .get()
+        ):
+            if category_ids := predicates.get("categoryPredicate"):
+                qs = Category.objects.filter(id__in=category_ids)
+                return create_connection_slice(
+                    qs, info, kwargs, CategoryCountableConnection
+                )
 
     @staticmethod
-    def resolve_channel_listings(root: ChannelContext[models.Sale], info: ResolveInfo):
-        return SaleChannelListingBySaleIdLoader(info.context).load(root.node.id)
+    def resolve_channel_listings(
+        root: ChannelContext[models.Promotion], info: ResolveInfo
+    ):
+        return SaleChannelListingByPromotionIdLoader(info.context).load(root.node.id)
 
     @staticmethod
     def resolve_collections(
-        root: ChannelContext[models.Sale], info: ResolveInfo, **kwargs
+        root: ChannelContext[models.Promotion], info: ResolveInfo, **kwargs
     ):
-        qs = root.node.collections.all()
-        qs = ChannelQsContext(qs=qs, channel_slug=root.channel_slug)
-        return create_connection_slice(qs, info, kwargs, CollectionCountableConnection)
+        if (
+            predicates := PredicateByPromotionIdLoader(info.context)
+            .load(root.node.id)
+            .get()
+        ):
+            if collection_ids := predicates.get("collectionPredicate"):
+                qs = Collection.objects.filter(id__in=collection_ids)
+                qs = ChannelQsContext(qs=qs, channel_slug=root.channel_slug)
+                return create_connection_slice(
+                    qs, info, kwargs, CollectionCountableConnection
+                )
```

Comment:
I think we should use then that will be calling the helper function like `return PredicateByPromotionIdLoader(info.context).load(root.node.id).then(_resolve_predicate)`.
I think that `get` might resolves the dataloader immediately. 
So IMO it should looks like:
```suggestion
       def _resolve_predicate(predicate):
            if not predicate:
                return None
            if collection_ids := predicates.get("collectionPredicate"):
                qs = Collection.objects.filter(id__in=collection_ids)
                qs = ChannelQsContext(qs=qs, channel_slug=root.channel_slug)
                return create_connection_slice(
                    qs, info, kwargs, CollectionCountableConnection
                )
        return PredicateByPromotionIdLoader(info.context).load(root.node.id).then(_resolve_predicate)
```

Reply:
Same for `products`, `variants`, `categories`

Reply:
Also, `PredicateByPromotionIdLoader` does not take channel into consideration but it should. IMO we should have datalaoder only for fetching promotion rule for given promotion and channel id, and use helper function for chaning predicate into different structure and then fetch the interesting data

Reply:
I agree with using `then`. However:

1. we can't fetch the data with `channel`, because it would be breaking change
2. I would also keep the "helper functionality" in dataloader, because this way we call it once instead of four times.